### PR TITLE
bootkube: Inject bootstrap MachineConfigs into cluster

### PR DESCRIPTION
**Pull request by cgwalters**
This is an ugly fix for
https://github.com/openshift/machine-config-operator/issues/367

### CODE REVIEW

File: data/data/bootstrap/files/usr/local/bin/bootkube.sh.template [hunk 1]
```diff
@@ -314,6 +314,17 @@ echo "etcd cluster up. Killing etcd certificate signer..."
...
+# Copy the bootstrap MCs to inject into the target cluster.  Ordinarily,
+# we don't need this.  If there is any drift between the bootstrap-generated
+# configs and the cluster-generated ones, it will incur a rollout (reboot)
+# of the cluster.  But, having that happen is way better than having the
+# MCO go degraded.  See https://github.com/openshift/machine-config-operator/issues/367
+for x in /etc/mcs/bootstrap/machine-configs/*.yaml; do
+    bn=$(basename $x)
+    (echo 'apiVersion: machineconfiguration.openshift.io/v1'
+     echo 'kind: MachineConfig'
+     python -c 'import sys,yaml; d=yaml.load(open(sys.argv[1])); del d["metadata"]["ownerReferences"]; yaml.dump(d, sys.stdout)' $x) > /opt/openshift/openshift/9a-${bn}
```
Comment (cgwalters) [inline data/data/bootstrap/files/usr/local/bin/bootkube.sh.template]: I forget now why I was doing this deletion of the owner references.  
Clearly what we should be doing is have the bootstrap process write out the API objects directly.

And we shouldn't depend on `python`.

**Comment by deads2k** [comment]
We have hit this in the field caused by a cluster-admin not getting all the options quite right. As I understand it, this can alleviate the "your cluster isn't installing" problem automatically in the unfortunate case and carries no negative side-effects in the common case. We should get this into a state where it (or something like it) can merge.

We've also hit it during development.  Basically, we aren't perfect and since we have the ability to make sure that the clusters start up, we should use it.

This doesn't necessarily obviate the need or benefit gotten from https://github.com/openshift/machine-config-operator/pull/1376 if we use it to provide a clear indication during installation status that the machine-config-operator isn't using the same content that was initially desired.  As I understand the current implications, right now the cluster is very difficult to recover cleanly because the actual content is missing.  Having the content available and choosing to fail with a clear "do this simple update to opt-in/fix/update" is a big win comparatively.

**Comment by cgwalters** [comment]
This overlaps with https://github.com/openshift/machine-config-operator/pull/1376 - which makes it easier to debug.  This PR would cause the system to automatically reconcile, and we're still debating whether that's a good idea or not.

For sure, if we merge this PR we need to add an e2e test that verifies each node in the cluster had exactly one reboot.

**Comment by deads2k** [comment]
> This overlaps with [openshift/machine-config-operator#1376](https://github.com/openshift/machine-config-operator/pull/1376) - which makes it easier to debug. This PR would cause the system to automatically reconcile, and we're still debating whether that's a good idea or not.

A re-reconcile doesn't necessarily mean that the operator can't go degraded on an unexpected rollout.  As I understand it, you can know that you are creating the "first" machineconfig because masters have known labels/names and you could annotate the machineconfig created during rendering.  If you did that, the MCC can see that only one master machineconfig exists and knows that it should get exactly the same result.  If it does not, it can still produce the expected output, but annotate in a way that say, "I've done this because I failed".  That signal could be picked up by the operator to report a failing status with a reconciled cluster and the cluster-admin gets a signal that something didn't go right, but he has a functional cluster to start recovering from instead of one reporting NotFound

**Comment by abhinavdahiya** [comment]
mc* is not allowed to reconcile multiple times during installation... It is a requirement because if during installation it is reconciling to multiple versions.. each would require a reboot. This is a special requirement for out mc* today. 

So imo we shouldn't think about solving this by pushing old rendered configuration... rather making sure we don't push ignition to masters unless all work from machineconfigs, kubeletconfig etc has completed

**Comment by smarterclayton** [comment]
> It is a requirement because if during installation it is reconciling to multiple versions.. each would require a reboot.

Can you describe what you mean here?

I thought the impact of this change was trying to capture 

a) installer desired state as rendered 
b) attempted calculated desired state from MCO once it boots

so that when we have bugs that show that delta we recognize it.

I agree that we should have 2 e2e tests - one that on install we verify we only reboot every node exactly once, and one that on upgrade every node is rebooted exactly once, and that violating those is a bug.  On an install however, it sounded like the argument was that in UPI when customers induce drift we need the extra info (installer gen vs MCO gen) to identify and highlight for the UPI user what their bug was.  I am definitely supportive of relaxing the hard rule in order to better identify the problem, and the e2e tests are a prerequisite to ensure we are actually following that (e2e tests are better at validating than we are)

**Comment by deads2k** [comment]
> I agree that we should have 2 e2e tests - one that on install we verify we only reboot every node exactly once, and one that on upgrade every node is rebooted exactly once, and that violating those is a bug.

@cgwalters @runcom I think these tests can be written against master.

**Comment by jstuever** [comment]
/uncc @jstuever 
This PR appears to have become stale.

**Comment by abhinavdahiya** [comment]
/close

closing due to inactivity.

**Pull request closed without merging**